compiler/incremental-compilation-impl/src/org/jetbrains/kotlin/incremental/classpathDiff/ClasspathSnapshotter.kt

```diff
@@ -57,26 +64,90 @@ object ClassSnapshotter {
 
     /** Creates [ClassSnapshot]s of the given classes. */
     fun snapshot(
-        classes: List<ClassFileWithContents>,
-        granularity: ClassSnapshotGranularity = CLASS_MEMBER_LEVEL,
+        classes: List<ClassFileWithContentsProvider>,
+        granularity: ClassSnapshotGranularity,
         metrics: BuildMetricsReporter = DoNothingBuildMetricsReporter
     ): List<ClassSnapshot> {
-        val classesInfo: List<BasicClassInfo> = metrics.measure(BuildTime.READ_CLASSES_BASIC_INFO) {
-            classes.map { it.classInfo }
+        fun ClassFile.getClassName(): JvmClassName {
+            check(unixStyleRelativePath.endsWith(".class", ignoreCase = true))
+            return JvmClassName.byInternalName(unixStyleRelativePath.dropLast(".class".length))
         }
-        val inaccessibleClassesInfo: Set<BasicClassInfo> = metrics.measure(BuildTime.FIND_INACCESSIBLE_CLASSES) {
-            findInaccessibleClasses(classesInfo)
+
+        val processedClasses = mutableSetOf<ClassId>()
+        val inaccessibleClasses = mutableSetOf<ClassId>()
+
+        // Returns `true` if this class is directly or transitively inaccessible, and `false` otherwise (or if we don't know).
+        // A class is transitively inaccessible if its outer class is (directly or transitively) inaccessible.
+        fun BasicClassInfo.isDirectlyOrTransitivelyInaccessible(): Boolean = when {
+            isInaccessible() -> true
+            else -> when (val outerClassId = classId.outerClassId) {
+                null -> false
+                in processedClasses -> outerClassId in inaccessibleClasses
+                else -> {
+                    // The outer class should have been processed due to the way we sorted the given classes (see the main algorithm where
+                    // this method is called). Here, we'll handle the uncommon case where the outer class is not yet processed (i.e., when
+                    // the outer class's path is not a prefix of this class's path).
+                    val outerClassName = JvmClassName.byClassId(outerClassId)
+                    // Don't optimize this search with indexing as it increases code complexity for a case that is unlikely to happen, and
+                    // we don't want to pay the indexing cost in the common case where indexing would not actually be used.
+                    val outerClass = classes.firstOrNull { it.classFile.getClassName() == outerClassName }
```

Comment:
I think we should be able to locate the entry directly, see my other comment?

I think we should either implement it efficiently, or just consider removing it entirely (i.e. we'll snapshot these classes).

Reply:
Ah you're right, I've finally found a way to implement it efficiently without adding complexity (it's actually simpler than the existing code in `master`), please take another look.